sdk2013: Mac: debug build fails with -Werror

**To reproduce:**
1. Clone into Metamod:Source
1. Clone the `sdk2013`
1. Build MM:S with `-s sdk2013` and `--enable-debug`

**Expected:**
Builds without error

**Actual:**
Build fails due to `-Werror`: (https://gist.github.com/maximsmol/bebf474bcb7c385ac6344bf5b5c84204)
```
error: add explicit braces to avoid dangling else [-Werror,-Wdangling-else]
                        DebuggerBreak();
                        ^
.../sdks/hl2sdk-sdk2013/public/tier0/platform.h:430:81: note: expanded from macro 'DebuggerBreak'
#define DebuggerBreak()  if ( Plat_IsInDebugSession() ) { __asm ( "int $3" ); } else { raise(SIGTRAP); }
```

If that is fixed as suggested in Potential Fix, fails with:
```
hl2sdk-sdk2013/public/tier1/utlblockmemory.h:340:6: error: unused variable 'nBlocksOld' [-Werror,-Wunused-variable]
        int nBlocksOld = m_nBlocks;
```

If that is fixed, fails in multiple locations with `-Wunused-local-typedef` resulting from `COMPILE_TIME_ASSERT`:
```
hl2sdk-sdk2013/public/tier0/threadtools.h:346:4: error: unused typedef 'generated_id_346' [-Werror,-Wunused-local-typedef]
                        COMPILE_TIME_ASSERT( sizeof(T) == sizeof(void *) );
                        ^
hl2sdk-sdk2013/public/tier0/dbg.h:570:50: note: expanded from macro 'COMPILE_TIME_ASSERT'
        #define COMPILE_TIME_ASSERT( pred ) typedef int UNIQUE_ID[ (pred) ? 1 : -1 ]
                                                        ^
hl2sdk-sdk2013/public/tier0/basetypes.h:372:19: note: expanded from macro 'UNIQUE_ID'
#define UNIQUE_ID UID_CAT2(UID_PREFIX,__LINE__)
                  ^
hl2sdk-sdk2013/public/tier0/basetypes.h:367:23: note: expanded from macro 'UID_CAT2'
#define UID_CAT2(a,c) UID_CAT1(a,c)
                      ^
hl2sdk-sdk2013/public/tier0/basetypes.h:366:23: note: expanded from macro 'UID_CAT1'
#define UID_CAT1(a,c) a ## c
                      ^
<scratch space>:16:1: note: expanded from here
generated_id_346
```

**Potential fix:**

`DebuggerBreak`:
Wrap macro in `do { ... } while (0)`

`nBlocksOld`:
```C++
int nBlockSize = NumElementsInBlock();
	int nBlocksOld = m_nBlocks;
	int nBlocks = ( numElements + nBlockSize - 1 ) / nBlockSize;

	// If the number of blocks is the same as the allocated number of blocks, we are done.
	if ( nBlocks == m_nBlocks )
		return;
```
`m_nBlocks` is clearly not modified between being saved to `nBlocksOld` and being compared to `nBlocks`, so this is likely not a bug and we can just remove `nBlocksOld`.

`COMPILE_TIME_ASSERT`:
Add `__attribute(unused)__` to the typedef as the code is conditional on GCC anyway.
```C++
#elif defined __GNUC__
	#define COMPILE_TIME_ASSERT( pred ) typedef int UNIQUE_ID[ (pred) ? 1 : -1 ]
```

--- a/public/tier0/dbg.h
+++ b/public/tier0/dbg.h
@@ -567,7 +567,7 @@ public:
 // of our complicated templates properly.  Use some preprocessor trickery
 // to workaround this
 #ifdef __GNUC__
-	#define COMPILE_TIME_ASSERT( pred ) typedef int UNIQUE_ID[ (pred) ? 1 : -1 ]
+	#define COMPILE_TIME_ASSERT( pred ) typedef int UNIQUE_ID[ (pred) ? 1 : -1 ] __attribute__((unused))
 #else
 	#if _MSC_VER >= 1600
 	// If available use static_assert instead of weird language tricks. This

--- a/public/tier0/dbg.h
+++ b/public/tier0/dbg.h
@@ -566,7 +566,15 @@ public:
 // We're using an ancient version of GCC that can't quite handle some
 // of our complicated templates properly.  Use some preprocessor trickery
 // to workaround this
-#ifdef __GNUC__
+#ifndef __has_feature
+  #define __has_feature(x) 0 // Compatibility with non-clang compilers.
+#endif
+#if __has_feature(cxx_static_assert)
+	// If available use static_assert instead of weird language tricks. This
+	// leads to much more readable messages when compile time assert constraints
+	// are violated.
+	#define COMPILE_TIME_ASSERT( pred ) static_assert( pred, "Compile time assert constraint is not true: " #pred )
+#elif defined __GNUC__
 	#define COMPILE_TIME_ASSERT( pred ) typedef int UNIQUE_ID[ (pred) ? 1 : -1 ] __attribute__((unused))
 #else
 	#if _MSC_VER >= 1600

--- a/public/tier0/platform.h
+++ b/public/tier0/platform.h
@@ -427,7 +427,7 @@ typedef void * HINSTANCE;
 	// On OSX, SIGTRAP doesn't really stop the thread cold when debugging.
 	// So if being debugged, use INT3 which is precise.
 #ifdef OSX
-#define DebuggerBreak()  if ( Plat_IsInDebugSession() ) { __asm ( "int $3" ); } else { raise(SIGTRAP); }
+#define DebuggerBreak()  do { if ( Plat_IsInDebugSession() ) { __asm ( "int $3" ); } else { raise(SIGTRAP); } } while (0)
 #else
 #define DebuggerBreak()  raise(SIGTRAP)
 #endif

--- a/public/tier1/utlblockmemory.h
+++ b/public/tier1/utlblockmemory.h
@@ -337,7 +337,6 @@ void CUtlBlockMemory<T,I>::Purge( int numElements )
 	}
 
 	int nBlockSize = NumElementsInBlock();
-	int nBlocksOld = m_nBlocks;
 	int nBlocks = ( numElements + nBlockSize - 1 ) / nBlockSize;
 
 	// If the number of blocks is the same as the allocated number of blocks, we are done.

